src/coreclr/tools/aot/crossgen2/Program.cs

```diff
@@ -253,6 +253,22 @@ public int Run()
                 }
             }
 
+            Dictionary<string, AssemblyNameInfo> allAssemblyNames = new ();
+            foreach (var rModule in _referenceableModules)
+            {
+                // Check for a loose match (two assemblies with the same name, i.e. 'System.Private.CoreLib')
+                var moduleName = rModule.Assembly.GetName();
+                if (allAssemblyNames.TryGetValue(moduleName.Name, out var firstModule))
+                {
+                    // Now check whether the full name is an exact match. If it's not, this indicates that the user accidentally provided two versions of the same
+                    //  assembly during compilation, and the results will probably not be what they want, so we should generate a warning message.
+                    if (firstModule.FullName != moduleName.FullName)
```

Comment:
Identical full names do not imply that assembly content is identical. It is very common for two assemblies to have identical full name but have different content. For example., assemblies from given nuget package that target different TFMs have identical full name.

If we want to change how we treat duplicate references, we should disallow or warn on any duplicated simple names, similar how the `strict` flag works for input assemblies.